google-cloud-tools-plugin/src/com/google/cloud/tools/intellij/appengine/sdk/ManagedCloudSdkService.java

```diff
@@ -16,35 +16,234 @@
+import com.google.cloud.tools.managedcloudsdk.install.SdkInstallerException;
+import com.google.cloud.tools.managedcloudsdk.install.UnknownArchiveTypeException;
+import com.google.common.annotations.VisibleForTesting;
+import com.google.common.collect.Lists;
+import com.google.common.util.concurrent.FutureCallback;
+import com.google.common.util.concurrent.Futures;
+import com.google.common.util.concurrent.ListenableFuture;
+import com.intellij.openapi.application.ApplicationManager;
+import com.intellij.openapi.diagnostic.Logger;
+import java.io.IOException;
 import java.nio.file.Path;
+import java.util.Collection;
 import org.jetbrains.annotations.Nullable;
 
 /**
  * {@link CloudSdkService} providing SDK that is managed and automatically installable, see {@link
- * com.google.cloud.tools.managedcloudsdk.ManagedCloudSdk}
+ * com.google.cloud.tools.managedcloudsdk.ManagedCloudSdk}.
+ *
+ * <p>Some methods ({@link #install()} do their work on background thread but listeners methods are
+ * always called on EDT thread. This class is not thread-safe, must be used on UI thread.
  */
-// TODO(ivanporty) implementation coming in the next PR
 public class ManagedCloudSdkService implements CloudSdkService {
+  private Logger logger = Logger.getInstance(ManagedCloudSdkService.class);
+
+  private ManagedCloudSdk managedCloudSdk;
+
+  private SdkStatus sdkStatus = SdkStatus.NOT_AVAILABLE;
+
+  private final Collection<SdkStatusUpdateListener> statusUpdateListeners = Lists.newArrayList();
+
+  private volatile ListenableFuture<Path> runningInstallationJob;
+
+  @Override
+  public void activate() {
+    // TODO track event that custom SDK is activated and used.
+
+    install();
+  }
 
   @Nullable
   @Override
   public Path getSdkHomePath() {
-    return null;
+    switch (sdkStatus) {
+      case READY:
+        return managedCloudSdk.getSdkHome();
+      default:
+        return null;
+    }
   }
 
   @Override
-  public void setSdkHomePath(String path) {}
+  public void setSdkHomePath(String path) {
+    /* unsupported, to be removed. */
+  }
 
   @Override
   public SdkStatus getStatus() {
-    return SdkStatus.NOT_AVAILABLE;
+    return sdkStatus;
   }
 
   @Override
-  public boolean installAutomatically() {
+  public boolean install() {
+    // check if installation is already running first, to prevent multiple jobs running.
+    if (runningInstallationJob == null || runningInstallationJob.isDone()) {
+      runningInstallationJob =
+          ThreadUtil.getInstance().executeInBackground(this::installManagedSdk);
+      Futures.addCallback(
+          runningInstallationJob,
+          new FutureCallback<Path>() {
+            @Override
+            public void onSuccess(Path path) {
+              if (path != null) {
+                logger.info("Managed Google Cloud SDK successfully installed at: " + path);
+              }
+            }
+
+            @Override
+            public void onFailure(Throwable t) {
```

Comment:
why not let the exceptions trickle up to this callback instead of catching them all in the install runnable?

Reply:
See my next comment please.